src/com/amazon/ion/impl/IonReaderBinaryIncremental.java

```diff
@@ -0,0 +1,1932 @@
+
+    // The index of the next byte to peek from the underlying buffer.
+    private int peekIndex = -1;
+
+    // ------
+
+    /**
+     * Constructor.
+     * @param inputStream the InputStream that provides binary Ion data.
+     */
+    IonReaderBinaryIncremental(InputStream inputStream) {
+        this(inputStream, EMPTY_CATALOG, null);
+    }
+
+    /**
+     * Constructor.
+     * @param inputStream the InputStream that provides binary Ion data.
+     * @param catalog the IonCatalog from which to resolve shared symbol table imports.
+     * @param configuration the configuration for the lookahead buffer.
+     */
+    IonReaderBinaryIncremental(
+        InputStream inputStream,
+        IonCatalog catalog,
+        IonBufferConfiguration configuration
+    ) {
+        this.inputStream = inputStream;
+        this.catalog = catalog;
+        if (configuration == null) {
+            lookahead = new IonReaderLookaheadBuffer(STANDARD_BUFFER_CONFIGURATION, inputStream);
+        } else {
+            lookahead = new IonReaderLookaheadBuffer(configuration, inputStream);
+        }
+        buffer = (ResizingPipedInputStream) lookahead.getPipe();
+        containerStack = new _Private_RecyclingStack<ContainerInfo>(
+            CONTAINER_STACK_INITIAL_CAPACITY,
+            CONTAINER_INFO_FACTORY
+        );
+        annotationSids = new ArrayList<Integer>(ANNOTATIONS_LIST_INITIAL_CAPACITY);
+        symbols = new ArrayList<String>(SYMBOLS_LIST_INITIAL_CAPACITY);
+        symbols.addAll(SYSTEM_SYMBOLS_1_0);
+        numberOfSymbols = SYSTEM_SYMBOLS_1_0_SIZE;
+        imports = new TreeMap<Integer, SymbolTable>();
+        scalarConverter = new _Private_ScalarConversions.ValueVariant();
+        resetImports();
+    }
+
+    /**
+     * Reusable iterator over the annotations on the current value.
+     */
+    private class AnnotationIterator implements Iterator<String> {
+
+        // All of the annotation SIDs on the current value.
+        private List<Integer> annotationSids = Collections.emptyList();
+        // The index into `annotationSids` containing the next annotation to be returned.
+        private int index = 0;
+
+        @Override
+        public boolean hasNext() {
+            return index < annotationSids.size();
+        }
+
+        @Override
+        public String next() {
+            int sid = annotationSids.get(index);
+            String annotation = getSymbol(sid);
+            if (annotation == null) {
+                throw new UnknownSymbolException(sid);
+            }
+            index++;
+            return annotation;
+        }
+
+        @Override
+        public void remove() {
+            throw new UnsupportedOperationException("This iterator does not support element removal.");
+        }
+
+        /**
+         * Reset the iterator so that it may be reused.
+         */
+        public void reset() {
+            index = 0;
+            annotationSids = getAnnotationSids();
+        }
+    }
```

Comment:
> I assume we're following this pattern to avoid allocating a new iterator?

Correct. This actually made a big difference in the benchmarks I ran that retrieved annotations on each value.

> This doesn't imply to me that future calls to iterateTypeAnnotations() can mutate another iterator I've retained, but I feel like I'm missing something.

No, that's a good callout. I don't expect anyone to be storing annotation iterators and iterating them after advancing the reader. That's what made me comfortable taking the risk. However, we should probably provide an option to enable/disable this behavior just in case. And, if we're going to have this more efficient way be default, that difference should be documented. I'll propose a new option on IonReaderBuilder.